keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    - mysql
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: wordpress-svc
+  labels:
+    app: wordpress
+    component: wordpress-svc
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  ports:
+    - port: 80
+  selector:
+    app: wordpress
+    component: wordpress-svc
+  type: LoadBalancer
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: wordpress-hsvc
+  labels:
+    app: wordpress
+    component: wordpress-hsvc
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  ports:
+    - port: 3306
+  selector:
+    app: wordpress
+    component: wordpress-hsvc
+  clusterIP: None
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: wordpress-webserver
+  labels:
+    app: wordpress
+    component: wordpress-webserver
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  replicas: 1
+  service: wordpress-hsvc
+  selector:
+    matchLabels:
+      app: wordpress
+      component: wordpress-webserver
+  template:
+    metadata:
+      labels:
+        app: wordpress
+        tier: frontend
+    spec:
+      containers:
+      - image: wordpress:4.8-apache
+        name: wordpress
+        env:
+        - name: WORDPRESS_DB_HOST
+          value: wordpress-mysql
+        - name: WORDPRESS_DB_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: mysql-pass
+              key: password
+        ports:
+        - containerPort: 80
+          name: wordpress
+        volumeMounts:
+        - name: wordpress-persistent-storage
+          mountPath: /var/www/html
+  volumeClaimTemplates:
+       - metadata:
+           name: mysql-persistent-storage
+         spec:
+           accessModes: [ "ReadWriteOnce" ]
+           resources:
+             requests:
+               storage: 250Gi
+```
+
+Above we generate two Applications, one for the MySQL RDBMs and another for the WordPress web server. Each of these 
+Applications can be managed independently. For instance, if the WordPress web server is deleted, the components of the 
+MySQL RDBMS will not be affected. However, the dependency between the two workloads is made explicit in the 
+`Spec.Dependencies` of the `wordpress` Application. Tools and UIs that examine the `wordpress` application are made 
+aware of its dependency on the `mysql` Application, and they can surface the dependency to users.
+
+#### Selecting Components
+In order to find the `Spec.Componenets` of an Application, a client should perform the following.
+
+1. For each GroupVersionKind in the Applications `Spec.Components`, list all objects of that Kind, in the namespace of 
+the Application, that match the Application's `Spec.Selector`.
+1. For each selected object, validate that the object has a `kubernetes.io/application` that matches the name of the 
+Application. Discard any objects that do not have such an annotation. This ensures that selector overlap does not 
+violate the intentions of the application creators and administrators.
```

Comment:
so the "component" label is not used...  not sure what the point of listing all the components in spec.components is, then.  you really only need to list the resource types.  (but i do think it would be better to explicitly select resources by a name, not just by type+label).